vllm/v1/executor/multiproc_executor.py

```diff
@@ -0,0 +1,168 @@
+
+        # Set multiprocessing envs that are common to V0 and V1
+        set_multiprocessing_worker_envs(self.parallel_config)
+
+        # Multiprocessing-based executor does not support multi-node setting.
+        # Since it only works for single node, we can use the loopback address
+        # 127.0.0.1 for communication.
+        distributed_init_method = get_distributed_init_method(
+            "127.0.0.1", get_open_port())
+
+        # Initialize worker and set up message queues for SchedulerOutputs
+        # and ModelRunnerOutputs
+        self.worker_request_mq = MessageQueue(self.world_size, self.world_size)
+        scheduler_output_handle = self.worker_request_mq.export_handle()
+
+        # Create workers
+        self.workers: List[WorkerProcHandle] = []
+        for rank in range(self.world_size):
+            worker = WorkerProc.make_worker_process(vllm_config, rank, rank,
+                                                    distributed_init_method,
+                                                    scheduler_output_handle)
+            self.workers.append(worker)
+
+        # Ensure message queues are ready. Will deadlock if re-ordered
+        # Must be kept consistent with the WorkerProc
+        self.worker_request_mq.wait_until_ready()
+        for w in self.workers:
+            w.worker_response_mq.wait_until_ready()
+
+    def initialize(self, num_gpu_blocks: int) -> None:
+        """
+        Initialize the KV caches and begin the model execution loop of the
+        underlying workers.
+        """
+        self.collective_rpc(
+            RPCParams(method="initialize_cache", output_ranks={}),
+            num_gpu_blocks)
+        self.collective_rpc(
+            RPCParams(method="compile_or_warm_up_model", output_ranks={}))
+
+    def determine_num_available_blocks(self) -> Tuple[int, int]:
+        """
+        Determine the number of available KV blocks by invoking the
+        underlying worker.
+        """
+        num_blocks = self.collective_rpc(
+            RPCParams(method="determine_num_available_blocks",
+                      output_ranks=set(range(self.world_size))))
+
+        # Since we use a shared centralized controller, we take the minimum
+        # number of blocks across all workers to make sure all the memory
+        # operators can be applied to all workers.
+        num_gpu_blocks = min(b[0] for b in num_blocks)
+        num_cpu_blocks = min(b[1] for b in num_blocks)
+
+        return num_gpu_blocks, num_cpu_blocks
+
+    def collective_rpc(self, rpc: RPCParams, *args, **kwargs) -> [Optional]:
+        """Execute a collective remote procedure call on workers"""
+        self.worker_request_mq.enqueue(
+            (rpc.method, rpc.output_ranks, args, kwargs))
+
+        try:
+            responses = [None] * self.world_size
+            for w in self.workers:
+                status, result = w.worker_response_mq.dequeue(
+                    timeout=rpc.timeout)
```

Comment:
nit: `timeout` should have a common starting time.

Reply:
oh yeah, this is a bug - nice catch!

Reply:
Switched the shm MessageQueue interface to take a deadline instead of a timeout